### ISSUE TITLE
ACP: Pattern methods for `OsStr` without `OsStr` patterns

### ISSUE THREAD

**Issue by epage** [rust-lang/libs-team#311]
# Proposal

## Problem statement

With rust-lang/rust#115443, developers, like those writing CLI parsers, can now perform (limited) operations on `OsStr` but it requires `unsafe` to get an `OsStr` back, requiring the developer to understand and follow some [very specific safety notes](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#method.from_encoded_bytes_unchecked) that cannot be checked by miri.

[RFC #2295](https://github.com/rust-lang/rfcs/pull/2295) exists for improving this but its been stalled out.  The assumption here is that part of the problem with that RFC is how wide its scope is and that by shrinking the scope, we can get some benefits now.

## Motivating examples or use cases

*Mostly copied from #306*

Argument parsers need to extract substrings from command line arguments. For example, `--option=somefilename` needs to be split into option and `somefilename`, and the original filename must be preserved without sanitizing it.

[`clap` currently implements](https://github.com/clap-rs/clap/blob/5c81df6855dbb12d1f1e28c7894aab6cc892abf3/clap_lex/src/ext.rs) `strip_prefix` and `split_once` using transmute (equivalent to the stable `encoded_bytes` APIs).

The [`os_str_bytes`](https://crates.io/crates/os_str_bytes) and [`osstrtools`](https://crates.io/crates/osstrtools) crates provides high-level string operations for OS strings. `os_str_bytes` is in the wild mainly used to convert between raw bytes and OS strings (e.g. [1](https://github.com/indygreg/linux-packaging-rs/blob/67643c0c604c0830a77a2f5794f831a75341322b/debian-packaging/src/deb/builder.rs#L10), [2](https://github.com/jonhoo/proximity-sort/blob/f31db09a3bf79f3b6166f920b4e71c73e750d431/src/main.rs#L67), [3](https://github.com/sharkdp/bat/blob/7658334645936d2a956fb19aa96e6fca849cb754/src/lessopen.rs#L115)). `osstrtools` enables reasonable uses of `split()` [to parse $PATH](https://github.com/rabite0/hunter/blob/355d9a3101f6d8dc375807de79e368602f1cb87d/src/minibuffer.rs#L379) and `replace()` [to fill in command line templates](https://github.com/rabite0/hunter/blob/355d9a3101f6d8dc375807de79e368602f1cb87d/src/proclist.rs#L63).

## Solution sketch

Provide  `str`s `Pattern`-accepting methods on `&OsStr`.

Defer out `OsStr` being used as a `Pattern` and `OsStr` indexing support which are specified in [RFC #2295](https://github.com/rust-lang/rfcs/pull/2295).

Example of methods to be added:
```rust
impl OsStr {
    pub fn contains<'a, P>(&'a self, pat: P) -> bool
    where
        P: Pattern<&'a Self>;

    pub fn starts_with<'a, P>(&'a self, pat: P) -> bool
    where
        P: Pattern<&'a Self>;

    pub fn ends_with<'a, P>(&'a self, pat: P) -> bool
    where
        P: Pattern<&'a Self>,
        P::Searcher: ReverseSearcher<&'a Self>;

    pub fn find<'a, P>(&'a self, pat: P) -> Option<usize>
    where
        P: Pattern<&'a Self>;

    pub fn rfind<'a, P>(&'a self, pat: P) -> Option<usize>
    where
        P: Pattern<&'a Self>,
        P::Searcher: ReverseSearcher<&'a Self>;

    // (Note: these should return a concrete iterator type instead of `impl Trait`.
    //  For ease of explanation the concrete type is not listed here.)
    pub fn split<'a, P>(&'a self, pat: P) -> impl Iterator<Item = &'a Self>
    where
        P: Pattern<&'a Self>;

    pub fn split_inclusive<'a, P>(&'a self, pat: P) -> impl Iterator<Item = &'a Self>
    where
        P: Pattern<&'a Self>;

    pub fn rsplit<'a, P>(&'a self, pat: P) -> impl Iterator<Item = &'a Self>
    where
        P: Pattern<&'a Self>,
        P::Searcher: ReverseSearcher<&'a Self>;

    pub fn split_terminator<'a, P>(&'a self, pat: P) -> impl Iterator<Item = &'a Self>
    where
        P: Pattern<&'a Self>;

    pub fn rsplit_terminator<'a, P>(&'a self, pat: P) -> impl Iterator<Item = &'a Self>
    where
        P: Pattern<&'a Self>,
        P::Searcher: ReverseSearcher<&'a Self>;

    pub fn splitn<'a, P>(&'a self, n: usize, pat: P) -> impl Iterator<Item = &'a Self>
    where
        P: Pattern<&'a Self>;

    pub fn rsplitn<'a, P>(&'a self, n: usize, pat: P) -> impl Iterator<Item = &'a Self>
    where
        P: Pattern<&'a Self>,
        P::Searcher: ReverseSearcher<&'a Self>;

    pub fn split_once<'a, P>(&'a self, delimiter: P) -> Option<(&'a Self, &'a Self)>where
        P: Pattern<&'a Self>;

    pub fn rsplit_once<'a, P>(&'a self, delimiter: P) -> Option<(&'a Self, &'a Self)>where
        P: Pattern<&'a Self>;

    pub fn matches<'a, P>(&'a self, pat: P) -> impl Iterator<Item = &'a Self>
    where
        P: Pattern<&'a Self>;

    pub fn rmatches<'a, P>(&self, pat: P) -> impl Iterator<Item = &'a Self>
    where
        P: Pattern<&'a Self>,
        P::Searcher: ReverseSearcher<&'a Self>;

    pub fn match_indices<'a, P>(&self, pat: P) -> impl Iterator<Item = (usize, &'a Self)>
    where
        P: Pattern<&'a Self>;

    pub fn rmatch_indices<'a, P>(&self, pat: P) -> impl Iterator<Item = (usize, &'a Self)>
    where
        P: Pattern<&'a Self>,
        P::Searcher: ReverseSearcher<&'a Self>;

    pub fn trim_matches<'a, P>(&'a self, pat: P) -> &'a Self
    where
        P: Pattern<&'a Self>,
        P::Searcher: DoubleEndedSearcher<&'a Self>;

    pub fn trim_start_matches<'a, P>(&'a self, pat: P) -> &'a Self
    where
        P: Pattern<&'a Self>;

    pub fn strip_prefix<'a, P>(&'a self, prefix: P) -> Option<&'a Self> where
    P: Pattern<&'a Self>;

    pub fn strip_suffix<'a, P>(&'a self, prefix: P) -> Option<&'a Self> where
    P: Pattern<&'a Self>;

    pub fn trim_end_matches<'a, P>(&'a self, pat: P) -> &'a Self
    where
        P: Pattern<&'a Self>,
        P::Searcher: ReverseSearcher<&'a Self>;

    pub fn replace<'a, P>(&'a self, from: P, to: &'a Self) -> Self::Owned
    where
        P: Pattern<&'a Self>;

    pub fn replacen<'a, P>(&'a self, from: P, to: &'a Self, count: usize) -> Self::Owned
    where
        P: Pattern<&'a Self>;
}

impl Pattern<&OsStr> for char {}
impl Pattern<&OsStr> for &str {}
impl Pattern<&OsStr> for &String {}
impl Pattern<&OsStr> for &[char] {}
impl Pattern<&OsStr> for &&str {}
impl<const N: usize> Pattern<&OsStr> for &[char; N] {}
impl<F: FnMut(char) -> bool> Pattern<&OsStr> for F {}
impl<const N: usize> Pattern<&OsStr> for [char; N] {}
```
- This is meant to match `str` and if there are any changes between the writing of this ACP and implementation, the focus should be on what `str` has at the time of implementation (e.g. not adding a deprecated variant but the new one)
- We likely want to add `trim`, `trim_start`, and `trim_end` to be consistent with `trim_start_matches` / `trim_end_matches`
- for more details, see rust-lang/rust#109350

This should work because
- rust-lang/rust#109698 already established that operations on UTF-8 / 7-bit ASCII boundaries are safe
- [It was decided to seal `Pattern`](https://github.com/rust-lang/rust/pull/76901#issuecomment-880169952) and, for now, `Pattern` is nightly only, allowing a lot of flexibility for how we implement `OsStr` support in the future (e.g. we could go as far as creating a `OsPattern` trait and switching to it without breaking anyone)

From an API design perspective, there is strong precedence for it
- Its copying methods over from `str`
- The design is a subset of [RFC #2295](https://github.com/rust-lang/rfcs/pull/2295) (approved) and [RFC #1309](https://github.com/rust-lang/rfcs/pull/1309) (postponed)
  - By deferring support for `OsStr` as a pattern, we bypass the main dividing point between proposals (split APIs, panic on unpaired surrogates, switching away from WTF-8)

## Alternatives

#306 proposes a `OsStr::slice_encoded_bytes`
- Still requires writing higher level operations on top, but at least its without `unsafe`
- Either takes a performance hit to be consistent across platforms or has per-platform caveats that will be similarly hard to get right for less common platforms among developers (e.g. Windows)
- As far as I can tell, there isn't precedence for an API design like this meaning more new ground has to be set (naming, deciding the above preconditions, etc)

## Links and related work

- #306
- #114
- [os_str_bytes](https://crates.io/crates/os_str_bytes)
- [osstrtools](https://crates.io/crates/osstrtools)
- [RFC #1309](https://github.com/rust-lang/rfcs/pull/1309)
- [RFC #2295](https://github.com/rust-lang/rfcs/blob/master/text/2295-os-str-pattern.md)
- [Decision to keep `Pattern` private](https://github.com/rust-lang/rust/pull/76901#issuecomment-880169952)
- https://github.com/rust-lang/rust/issues/111544
- https://github.com/rust-lang/libs-team/issues/114
- https://github.com/rust-lang/rust/pull/109350

## What happens now?

This issue contains an API change proposal (or ACP) and is part of the libs-api team [feature lifecycle]. Once this issue is filed, the libs-api team will review open proposals as capability becomes available. Current response times do not have a clear estimate, but may be up to several months.

[feature lifecycle]: https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html

## Possible responses

The libs team may respond in various different ways. First, the team will consider the *problem* (this doesn't require any concrete solution or alternatives to have been proposed):

- We think this problem seems worth solving, and the standard library might be the right place to solve it.
- We think that this probably doesn't belong in the standard library.

Second, if there's a concrete solution:

- We think this specific solution looks roughly right, approved, you or someone else should implement this. (Further review will still happen on the subsequent implementation PR.)
- We're not sure this is the right solution, and the alternatives or other materials don't give us enough information to be sure about that. Here are some questions we have that aren't answered, or rough ideas about alternatives we'd want to see discussed.

**Comment by joshtriplett** [comment]
This seems like a good idea to me. Putting these methods on OsStr will allow code to do simple parsing/splitting/etc in safe code. And because this does *not* add OsStr to Pattern, it remains a simple addition to API surface without any representation changes.

**Comment by joshtriplett** [comment]
@pitaj Presumably we'd do this the same way we already did for today's patterns.

**Comment by pitaj** [comment]
AFAIK all of today's `Pattern`s are defined in core, but `OsStr` lives in std.

**Comment by joshtriplett** [comment]
@pitaj Ah, I see. I *think* we have a mechanism that allows us to have impls of core traits in std without regard for the orphan rule, which would address that.

**Comment by pitaj** [comment]
Last I checked, there's a mechanism for inherent impls, but not for traits.

**Comment by programmerjake** [comment]
we could put just the `OsStr` struct in `core` as an implementation detail, along with necessary trait impls, and put the rest in `std` with a re-export of `OsStr`...this would allow implementing:
> ```rust
> impl<F: FnMut(char) -> bool> Pattern<&OsStr> for F {}
> ```

**Comment by m-ou-se** [comment]
We briefly discussed this in last week's libs-api meeting. While we agree it'd be good for OsStr to have a more complete API, we're worried about the amount of string types: should CStr and `[ascii::Char]` have the same api, for example?

It'd be good to first explore solutions that could benefit all string types, before continuing with this proposal to extend OsStr itself. For example, could a trait or another mechanism be used to make this api availabel to all string types?

**Comment by epage** [comment]
So if I understand correctly, the desire is to explore the design of a Pattern Extension trait with methods like `contains`, `starts_with`, etc that can apply to all string-like types?

Including `CStr` in this list is interesting because some methods, like `strip_prefix`, will require allocating into a `CString`.  To be clear, that is expected and part of the design requirements for this?

**Comment by programmerjake** [comment]
> Including `CStr` in this list is interesting because some methods, like `strip_prefix`, will require allocating into a `CString`.

i think you meant `strip_suffix`? `strip_prefix` should be able to return `&CStr`

**Comment by epage** [comment]
Good point, and figuring out how we should handle `strip_prefix` vs all other slicing operations adds another area for design exploration and discussions.  Do we have all slicing operations treated the same, having `CStr` always return an owned type?  Or do we hard code into the trait that only leading/middle/arbitrary content might be owned and trailing content is always borrowed.

For example, `CStr::split_once` could return either `(Owned, &Borrowed)` or `(Owned, Owned)` while `split` would always return `impl Iterator<Item=Owned>`.

**Comment by programmerjake** [comment]
maybe `CStr::split_once` could return something more like `(&[u8], &CStr)`?

**Comment by blyxxyz** [comment]
`CStr` operations could return `&[u8]` and let the caller decide whether to turn it into `CString` (at the cost of a null byte check). Or they could mutate the original string to insert null bytes, `strtok`-style (probably a bad idea). Or there could be a `CStrSlice` type that has no terminator but is guaranteed to contain no null bytes...

Should function arguments be full-blown `CStr`s? It's important that they not have internal null bytes but for e.g. a search pattern there's no technical reason for the terminator, and in some cases that could require an allocation. (The recently stabilized C string literals would help with this at least.)

But mainly: Should `CStr` have this API at all? My experience is very limited but it does include a [case](https://github.com/uutils/coreutils/commit/f2ddae93fad3bd0e93da5c19300a23befec59221#diff-14e34feabaab900738e1f95ce5121d56227ae86c3e68b898e2a4e4ff7623e1ed) where raw C string pointers were passed around arbitrarily when they should have been converted ASAP after the function call. How big of a use case is there for manipulating these instead of converting them immediately at the FFI boundary? Unlike OS strings they're just bytes with a validity requirement, so they're easier to convert.

Should `[u8]` get this API? `Cstr` would then get lossless access through `to_bytes()`.

**Comment by programmerjake** [comment]
wouldn't `CStrSlice` just be `[NonZeroU8]` (except that's kinda hard to work with currently due to lacking lots of byte string methods)

**Comment by tgross35** [comment]
I like the trait idea, especially since it will allow writing combinators that work for any string type and being able to reuse them on both `CStr`s from FFI and `&str`s from users.

Maybe a trait could look something like this (ignoring lifetimes):

```rust
trait SliceLikePattern: ToOwned {
    // Yes, we don't have associated type defaults...

    /// Result of splitting items
    type Partial = Self;

    /// Rightmost result of split items if different than `Partial`, e.g. for `CStr`
    type PartialRight = Self::Partial;

    /// Pattern type used when a single element will be extracted. `u8` for `&[u8]`,
    /// `str::pattern::Pattern` for str, maybe `u8` or `u16` for `OsStr`
    /// Or maybe  `FnMut(&u8) -> bool` for slices, as in `split_once`
    type ItemPattern;

    /// Pattern type used when a partial (slice) is expected, `&[u8]` for `&[u8]`
    /// still `str::pattern::Pattern` for `str`
    type PartialPattern;

    /// PartialPattern but if there is a specific right-first search
    /// e.g. str's `<P as Pattern<'a>>::Searcher: ReverseSearcher<'a>`
    type PartialPatternReverse = Self::PartialPattern;
    
    fn split_at(&self, mid: usize) -> (&Self::Partial, &Self::PartialRight);
    fn split_at_mut(&self, mid: usize) -> (&mut Self::Partial, &mut Self::PartialRight);
    fn contains<P: Self::ItemPattern>(&self, pat: P) -> bool;
    fn starts_with<P: Self::PartialPattern>(&self, pat: P) -> bool;
    fn ends_with<P: Self::PartialPatternReverse>(&self, pat: P) -> bool;
    fn find<P: Self::PartialPattern>(&self, pat: P) -> Option<usize>;
    fn rfind<P: Self::PartialPatternReverse>(&self, pat: P) -> Option<usize>;
    fn split<P: Self::PartialPattern>(&self, pat: P) -> Split<P>;
    // ... similar variants of iterating splits and matches
    fn split_once<P: Self::ItemPattern>(&self, pat: P) -> Option<(&Self::Partial, &Self::PartialRight)>;
    fn rsplit_once<P: Self::ItemPatternReverse>(&self, pat: P) -> Option<(&Self::Partial, &Self::PartialRight)>;
    // I don't think we can do simple `trim_{start, end}` here or anything else that
    // relies on whitespace knowledge
    fn trim_start_matches<P: Self::PartialPattern>(&self, pat: P) -> &Self::PartialRight;
    fn trim_end_matches<P: Self::PartialPatternReverse>(&self, pat: P) -> &Self::Partial;
    fn strip_prefix<P: Self::PartialPattern>(&self, pat: P) -> Option<&Self::PartialRight>;
    fn strip_suffix<P: Self::PartialPatternReverse>(&self, pat: P) -> Option<&Self::Partial>;
    fn replace<P: Self::PartialPattern>(&'a self, from: P, to: &Self::PartialRight) -> <Self as ToOwned>::ToOwned;
    fn repeat<P: Self::PartialPattern>(&'a self, from: P, repeat: usize) -> <Self as ToOwned>::ToOwned;
}
```

There probably isn't anything that restricts this to string-like types, I could see a lot of this being beneficial to let this apply to anything.

**Comment by joshtriplett** [comment]
We discussed this one again in today's @rust-lang/libs-api meeting, in light of https://github.com/rust-lang/libs-team/issues/499.

We'd like to accept this, using the same API proposal as https://github.com/rust-lang/libs-team/issues/499: use the same `BytePattern` type, make sure that `BytePattern` is implemented for `u8`/`[u8; N]`/`&[u8]`/`Vec<u8>``OsStr`/`OsString`/`Path`/`PathBuf`/`str`/`String`, and use the same `*_bytes` methods being proposed there.

**Comment by epage** [comment]
@joshtriplett from my understanding of #499, we'd have functionality like
```rust
trait ByteSlice {
    fn split_bytes(&self, pat: impl BytePattern) -> BytesSplit<'_, P>;
}

trait BytePattern { ... }

impl BytePattern for &[u8] { ... }

// `BytesSplit<'_, P>` would be an `Iterator<&[u8]>`
```

If that is correct, that misses the key goal of this ACP: use of `OsStr` without `unsafe`.  Safe manipulation of `OsStr` requires patterns only be UTF-8 and that the output type is `OsStr` (we could add checked constructors for `OsStr` but that is extra work that isn't needed in these cases and has its own design decisions to work out).

Use of a `ByteSlice` trait would require `unsafe` to turn the result back into an `OsStr` and a `// SAFETY` requirement about what types or values are used for `BytePattern`.

If we limit `BytePattern` to 7-bit ASCII or UTF-8 and have an associated type for the return value so `OsStr` can preserve itself, then we could avoid `unsafe`.

**Comment by Amanieu** [comment]
No, what is being proposed is that we have inherent methods on `OsStr` that return `OsStr` slices, but that the pattern accepted for matching can be any sequence of bytes (via the `BytePattern` trait).

**Comment by blyxxyz** [comment]
I don't think it's sound to accept arbitrary byte patterns for Windows/WTF-8 OS strings. Consider:

```rust
let crab = std::ffi::OsStr::new("🦀");
assert_eq!(crab.as_encoded_bytes(), b"\xF0\x9F\xA6\x80");
let (head, tail) = crab.split_once(b"\x9F").unwrap();
```

`head` would now be `"\xF0"`, but that's not valid WTF-8, so it causes UB and it can't be transformed into UTF-16.

- Accepting `OsStr` patterns is (I think) the most permissive API that's still safe and sensible.
- Accepting only `str` patterns like this ACP proposes is slightly less permissive but much easier to implement while still providing almost all the value.
- Accepting `[u8]` patterns would be too permissive. (Unless you reject/ignore problematic patterns at runtime, but that seems worse than restricting the patterns in the type system.)

**Comment by epage** [comment]
In that case, I assume we'd have slicing logic like #306.

**Comment by joshtriplett** [comment]
@epage Thanks for drawing attention to this; I think we missed that. Clarifying the issue here: the set of methods isn't a problem here, it's just a question of what types they accept and return, right? It makes sense that BytePattern may be too general here.

**Comment by epage** [comment]
Agreed

Some options for `BytePattern`:
- Mark the functions as `unsafe`
- Require UTF-8 boundaries (cross-platform expectations) and
  - panic if its not
  - fine some sentinel answer
- Require platform-specific bounadires and the same above ways of responding

We could also go back to the original proposal and use `Pattern` and skip over the non-UTF8 parts.  We could still offer the `BytePattern` variants and return `&[u8]`.

**Comment by joshtriplett** [comment]
I definitely don't think we'll want to mark the functions `unsafe`. That would lose much of the value of providing them.

Of the various solutions, which path would you recommend, balancing simplicity, usability, and consistency with #499?

**Comment by tgross35** [comment]
Could https://github.com/rust-lang/libs-team/issues/311#issuecomment-1882102190 (or a simplified version of it) work? This would provide methods similar to those from `BytePattern`, with the exception that the return values are associated types.

**Comment by joshtriplett** [comment]
@epage Got it! So, use `Pattern` to restrict the *input* to UTF-8, and rely on `OsStr`/`OsString` being supersets of UTF-8?

**Comment by epage** [comment]
Yes, to restrict the patterns/needles to UTF-8 while the haystack can be anything.  Non-UTF-8 content won't match.  There might be some details with that to work out but i figure thats what the unstable period will be for.

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/libs-api meeting.

We realized that using `Pattern` would be a problem because of the `FnMut(char) -> bool` case.

We considered three alternatives:

Option A: use Pattern and call the function with a Unicode replacement character for non-UTF-8 (using the logic from https://doc.rust-lang.org/std/str/struct.Utf8Error.html for how many bytes to treat as the invalid character)

Option B: `AsRef<OsStr>`.

Option C: A sealed `OsStrPattern` trait, which initially has all the variants of `Pattern` *except* `FnMut(char) -> bool`.

We decided on option C for the initial experiment, and we can evaluate whether this was the right choice at stabilization time.

**Comment by epage** [comment]
> We realized that using Pattern would be a problem because of the FnMut(char) -> bool case.

For the record, the `FnMut` problem is that it requires turning parts of the `OsStr` to a `char` and you can do `|c| c != '/'`.  In doing a negative match, we need to do *something* regarding the non-UTF8 data.

**Comment by clarfonthey** [comment]
Going to count this as being tracked by rust-lang/rust#49802.
